File: 142_Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Let a = head to cycle start, b = head to meet, c = cycle length.
        fast = b + kc = 2 * slow = 2 * b, so b = kc.

        For the slow ptr moves b-a step in the cycle, it will complete cycles and reach the cycle start by moving a steps.
        """
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                while head != slow:
                    head = head.next
                    slow = slow.next
                return head
        return None

File: 142_Linked List Cycle II.java
// Time complexity: O(n)
// Space complexity: O(1)

// head -> cycle entry: distance a
// cycle entry -> fast/slow ptr meet: distance b
// fast/slow ptr meet -> cycle entry: distance c
// therefore, cycle length = b + c

// case 1: when fast / slow ptr meet
// fast moves: a + b + (b + c)*k, k is an integer
// slow moves: a + b
// Given the fast ptr moves twice as the slow ptr: 
//      2*(a + b) = a + b + (b + c)*k
//      2*(a + b) - (a + b) - (b + c) = (b + c)*(k - 1)
//      a - c = (b + c)*(k - 1)

// case 2: have a new ptr to head after they meet, then have both start to move
// when slow reaches the cycle entry, the new ptr is (a-c) away from the entry, which is proportional to the cycle length
// therefore, if both keep iterating, they will meet at the cycle entry

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head, slow = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;

            if (fast == slow) {
                while (head != slow) {
                    slow = slow.next;
                    head = head.next;
                }
                return head;
            }
        }

        return null;
    }
}